### PR TITLE
CLOUDSTACK-8609: [VMware] VM is not accessible after migration across clusters

### PR DESCRIPTION
**Pull request by sureshanaparti**
[VMware] VM is not accessible after migration across clusters.

- Once a VM is successfully started, don't delete the files associated with the unregistered VM, if the files are in a storage that is being used by the new VM.
- Attempt to unregister a VM in another DC, only if there is a host associated with a VM.

This closes #556

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@borisstoyanov Can you please run VMware trillian tests on this PR

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by Divyanimmagadda** [comment]
Test verified. LGTM.

**Comment by Divyanimmagadda** [comment]
Fix verified. LGTM.

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 842
 Hypervisor xenserver
 NetworkType Advanced
 Passed=109
 Failed=4
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 30 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 30 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 30 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 30 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by rhtyd** [comment]
@sureshanaparti can you rebase this against latest 4.11.
I'll in the interim move this to 4.11.2.0 milestone, if the author does not repond I may take over this PR in future.

**Comment by sureshanaparti** [comment]
@rhtyd I'll rebase and update. Thanks.

**Comment by rhtyd** [comment]
@blueorangutan package

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 1]
```diff
@@ -4905,7 +4924,7 @@ protected Answer execute(UnregisterVMCommand cmd) {
                     VirtualMachineFileLayoutEx vmFileLayout = vmMo.getFileLayout();
                     context.getService().unregisterVM(vmMo.getMor());
                     if (cmd.getCleanupVmFiles()) {
-                        deleteUnregisteredVmFiles(vmFileLayout, dataCenterMo, false);
+                        deleteUnregisteredVmFiles(vmFileLayout, dataCenterMo, false, null);
```
Comment (rafaelweingartner) [inline plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: So, if there is an exception, you do delete the files, even if you shouldn't?

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java [hunk 2]
```diff
@@ -932,6 +933,38 @@ else if (prop.getName().startsWith("value[")) {
         return networks;
     }
 
+    public List<DatastoreMO> getAllDatastores() throws Exception {
```
Comment (rafaelweingartner) [inline vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java]: This "Exception" here. Is it from a VMware JDK method?

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 3]
```diff
@@ -2253,7 +2255,18 @@ protected StartAnswer execute(StartCommand cmd) {
 
             // Since VM was successfully powered-on, if there was an existing VM in a different cluster that was unregistered, delete all the files associated with it.
             if (existingVmName != null && existingVmFileLayout != null) {
```
Comment (rafaelweingartner) [inline plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Can you extract this IF body to a method?
This `execute` method is huge.

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 3 above)
Comment (sureshanaparti) [reply plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: @rafaelweingartner Sure. Will update.

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java (hunk 2 above)
Comment (sureshanaparti) [reply vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java]: @rafaelweingartner Correct. VMware SDK exception.

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 4]
```diff
@@ -2954,23 +2967,29 @@ private void deleteUnregisteredVmFiles(VirtualMachineFileLayoutEx vmFileLayout,
...
+                        s_logger.debug("Deleting file: " + vmDiskFileFullPath);
+                        vmFolderDsMo.deleteFile(vmDiskFileFullPath, dcMo.getMor(), true);
+                    }
                 }
             }
             // Delete VM folder
             if (deleteDisks || files.length == 0) {
-                s_logger.debug("Deleting folder: " + vmFolder.getPath());
-                vmFolderDsMo.deleteFolder(vmFolder.getPath(), dcMo.getMor());
+                if (skipDatastores == null || !skipDatastores.contains(vmFolderDsMo.getName())) {
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: there is some repetition here. maybe worth extracting into a method. (even though two are for files and one for a folder)

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java (hunk 2 above)
Comment (DaanHoogland) [reply vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java]: @sureshanaparti is it worth handling it locally and/or throwing a more specific exception up?

**Comment by PaulAngus** [comment]
@sureshanaparti are you still working on this? Could you address the comments please,

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 4 above)
Comment (sureshanaparti) [reply plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Will check and update. thanks for the review @DaanHoogland

**Comment by sureshanaparti** [comment]
Yes @PaulAngus. I'll address them. Thanks!

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java (hunk 2 above)
Comment (GabrielBrascher) [reply vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java]: IMHO, the code would benefit from a more specific exception.

**Comment by DaanHoogland** [comment]
@sureshanaparti If you have the code that addresses the comments lying around, this is the time to update. We may be able to test before RC3 for 4.11.1

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java (hunk 2 above)
Comment (rafaelweingartner) [reply vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java]: @GabrielBrascher it is not possible. The methods from VMware SDK are throwing `Exception`.

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 3 above)
Comment (rafaelweingartner) [reply plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Perhaps you forgot to push the commit?

**Comment by rhtyd** [comment]
@sureshanaparti ping, can you address outstanding comments.
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan help

**Comment by DaanHoogland** [comment]
@blueorangutan test centos6 vmware-55u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65

**Comment by rhtyd** [comment]
Review request - @DaanHoogland @rafaelweingartner @resmo and others

**Comment by DaanHoogland** [comment]
@blueorangutan test centos6 vmware-55u3

**Comment by DaanHoogland** [comment]
@blueorangutan test centos6 vmware-55u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-55u3

**Comment by DaanHoogland** [comment]
is scsi virtio supposed to work (in vmware)? @rhtyd @rafaelweingartner, if not, any other test needed?

**Comment by borisstoyanov** [comment]
screen shot 2018-08-21 at 14 04 06
Looks like it's properly skipping all non kvm hypervisors

**Comment by rhtyd** [comment]
@borisstoyanov looks like there was an issue in the tear down of that test

**Comment by rhtyd** [comment]
Merging this based on test results and reviews

**Pull request merged**